Android FreeRDP: a single tap is a single left click

**Is your feature request related to a problem? Please describe.**
Right now the issue I'm facing is that, for a single left click on the host, I have to double tap. For example, in Microsoft's RDP the user only needs to single tap. Don't know if single tapping right now is used for another feature.
For example if in the host I have to double click on something, I don't need to tap 4 times (2 for each single tap, and this doesn't work all the times and you find yourself spamming taps at the screen).
If this is the case, maybe an option to disable

**Describe the solution you'd like**
When I tap a single time, is registered on the host as a left click.

**Describe alternatives you've considered**
An option to disable the current single tap feature (don't know what is that used for).

**Additional context**
My plan is to use aFreeRDP as a better alternative to Microsoft's RDP client to use it on CK65 Android scanners at my company. Right now, Microsoft RDP client is in a bugged state, the UI is unusable for the workers and the performance is awful compared to old Windows embedded 6.5 scanners (intermec ck3).
For the tests that I have done, aFreeRDP is vastly better in our case scenario, the only drawback is to double tap for a single click, which confuses the workers.

I've been trying to build the apk from the code available, so I can start working on a fix for myself at least, but I came to lot of errors and such, so I needed to modify some parts of the gradle config files to get started (I'm on Windows because it's my work's laptop). Right now, when I do gradlew build I get errors:
`> Build command failed.
  Error while executing process C:\Users\cergutiv\AppData\Local\Android\Sdk\cmake\3.18.1\bin\ninja.exe with arguments {-C C:\Users\cergutiv\StudioProjects\FreeRDP\client\Android\Studio\freeRDPCore\.cxx\Debug\44i2q3c1\arm64-v8a freerdp-android}
  ninja: Entering directory `C:\Users\cergutiv\StudioProjects\FreeRDP\client\Android\Studio\freeRDPCore\.cxx\Debug\44i2q3c1\arm64-v8a'
  [1/1] Linking C shared library C:\Users\cergutiv\StudioProjects\FreeRDP\client\Android\Studio\freeRDPCore\build\intermediates\cxx\Debug\44i2q3c1\obj\arm64-v8a\libfreerdp-android.so
  FAILED: C:/Users/cergutiv/StudioProjects/FreeRDP/client/Android/Studio/freeRDPCore/build/intermediates/cxx/Debug/44i2q3c1/obj/arm64-v8a/libfreerdp-android.so
  cmd.exe /C "cd . && C:\Users\cergutiv\AppData\Local\Android\Sdk\ndk\21.4.7075529\toolchains\llvm\prebuilt\windows-x86_64\bin\clang.exe --target=aarch64-none-linux-android23 --gcc-toolchain=C:/Users/cergutiv/AppData/Local/Android/Sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/windows-x86_64 --sysroot=C:/Users/cergutiv/AppData/Local/Android/Sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/windows-x86_64/sysroot -fPIC -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security  -O0 -fno-limit-debug-info  -Wl,--exclude-libs,libgcc.a -Wl,--exclude-libs,libgcc_real.a -Wl,--exclude-libs,libatomic.a -static-libstdc++ -Wl,--build-id -Wl,--fatal-warnings -Wl,--no-undefined -Qunused-arguments -shared -Wl,-soname,libfreerdp-android.so -o C:\Users\cergutiv\StudioProjects\FreeRDP\client\Android\Studio\freeRDPCore\build\intermediates\cxx\Debug\44i2q3c1\obj\arm64-v8a\libfreerdp-android.so CMakeFiles/freerdp-android.dir/android_event.c.o CMakeFiles/freerdp-android.dir/android_freerdp.c.o CMakeFiles/freerdp-android.dir/android_jni_utils.c.o CMakeFiles/freerdp-android.dir/android_jni_callback.c.o CMakeFiles/freerdp-android.dir/android_cliprdr.c.o  C:/Users/cergutiv/StudioProjects/FreeRDP/client/Android/Studio/freeRDPCore/src/main/jniLibs/arm64-v8a/libfreerdp3.so  C:/Users/cergutiv/StudioProjects/FreeRDP/client/Android/Studio/freeRDPCore/src/main/jniLibs/arm64-v8a/libfreerdp-client3.so  C:/Users/cergutiv/StudioProjects/FreeRDP/client/Android/Studio/freeRDPCore/src/main/jniLibs/arm64-v8a/libwinpr3.so  -llog  C:/Users/cergutiv/AppData/Local/Android/Sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/windows-x86_64/sysroot/usr/lib/aarch64-linux-android/libdl.a  -ljnigraphics  -latomic -lm && cd ."
  CMakeFiles/freerdp-android.dir/android_event.c.o: In function `android_process_event':
  C:/Users/cergutiv/StudioProjects/FreeRDP/client/Android/Studio/freeRDPCore/src/main/cpp/android_event.c:(.text.android_process_event+0xc): undefined reference to `assert'
  C:/Users/cergutiv/StudioProjects/FreeRDP/client/Android/Studio/freeRDPCore/src/main/cpp/android_event.c:(.text.android_process_event+0x10): undefined reference to `assert'
  CMakeFiles/freerdp-android.dir/android_freerdp.c.o: In function `RdpClientEntry':
  C:/Users/cergutiv/StudioProjects/FreeRDP/client/Android/Studio/freeRDPCore/src/main/cpp/android_freerdp.c:(.text.RdpClientEntry+0xc): undefined reference to `assert'
  C:/Users/cergutiv/StudioProjects/FreeRDP/client/Android/Studio/freeRDPCore/src/main/cpp/android_freerdp.c:(.text.RdpClientEntry+0x10): undefined reference to `assert'
  CMakeFiles/freerdp-android.dir/android_freerdp.c.o: In function `android_thread_func':
  C:/Users/cergutiv/StudioProjects/FreeRDP/client/Android/Studio/freeRDPCore/src/main/cpp/android_freerdp.c:(.text.android_thread_func+0xf4): undefined reference to `assert'
  CMakeFiles/freerdp-android.dir/android_freerdp.c.o:C:/Users/cergutiv/StudioProjects/FreeRDP/client/Android/Studio/freeRDPCore/src/main/cpp/android_freerdp.c:(.text.android_thread_func+0xf8): more undefined references to `assert' follow
  clang: error: linker command failed with exit code 1 (use -v to see invocation)
  ninja: build stopped: subcommand failed.`

Is this because I'm trying to build the package on Windows?

--- a/client/Android/Studio/freeRDPCore/src/main/java/com/freerdp/freerdpcore/presentation/SessionActivity.java
+++ b/client/Android/Studio/freeRDPCore/src/main/java/com/freerdp/freerdpcore/presentation/SessionActivity.java
@@ -1179,8 +1179,9 @@ public class SessionActivity extends AppCompatActivity
 		                           Mouse.getLeftButtonEvent(this, down));
 	}
 
-		if (!down)
-			toggleMouseButtons = false;
+	@Override public void onSessionViewMiddleTouch(int x, int y, boolean down)
+	{
+		LibFreeRDP.sendCursorEvent(session.getInstance(), x, y, Mouse.getMiddleButtonEvent(down));
 	}
 
 	@Override public void onSessionViewRightTouch(int x, int y, boolean down)
@@ -1202,6 +1203,11 @@ public class SessionActivity extends AppCompatActivity
 	@Override public void onSessionViewScroll(boolean down)
 	{
 		LibFreeRDP.sendCursorEvent(session.getInstance(), 0, 0, Mouse.getScrollEvent(this, down));
+	}
+
+	@Override public void onSessionViewHScroll(boolean right)
+	{
+		LibFreeRDP.sendCursorEvent(session.getInstance(), 0, 0, Mouse.getHScrollEvent(this, right));
 	}
 
 	// ****************************************************************************

--- a/client/Android/Studio/freeRDPCore/src/main/java/com/freerdp/freerdpcore/presentation/SessionActivity.java
+++ b/client/Android/Studio/freeRDPCore/src/main/java/com/freerdp/freerdpcore/presentation/SessionActivity.java
@@ -118,7 +118,6 @@ public class SessionActivity extends AppCompatActivity
 
 	private boolean connectCancelledByUser = false;
 	private boolean sessionRunning = false;
-	private boolean toggleMouseButtons = false;
 	private long backPressedTime = 0;
 
 	private LibFreeRDPBroadcastReceiver libFreeRDPBroadcastReceiver;
@@ -334,8 +333,6 @@ public class SessionActivity extends AppCompatActivity
 				zoomControls.setIsZoomInEnabled(true);
 			}
 		});
-
-		toggleMouseButtons = false;
 
 		createDialogs();
 
@@ -1179,22 +1176,27 @@ public class SessionActivity extends AppCompatActivity
 			cancelDelayedMoveEvent();
 
 		LibFreeRDP.sendCursorEvent(session.getInstance(), x, y,
-		                           toggleMouseButtons ? Mouse.getRightButtonEvent(this, down)
-		                                              : Mouse.getLeftButtonEvent(this, down));
+		                           Mouse.getLeftButtonEvent(this, down));
+	}
 
 		if (!down)
 			toggleMouseButtons = false;
 	}
 
-	public void onSessionViewRightTouch(int x, int y, boolean down)
+	@Override public void onSessionViewRightTouch(int x, int y, boolean down)
 	{
-		if (!down)
-			toggleMouseButtons = !toggleMouseButtons;
+		LibFreeRDP.sendCursorEvent(session.getInstance(), x, y,
+		                           Mouse.getRightButtonEvent(this, down));
 	}
 
 	@Override public void onSessionViewMove(int x, int y)
 	{
 		sendDelayedMoveEvent(x, y);
+	}
+
+	@Override public void onSessionViewMouseMove(int x, int y)
+	{
+		LibFreeRDP.sendCursorEvent(session.getInstance(), x, y, Mouse.getMoveEvent());
 	}
 
 	@Override public void onSessionViewScroll(boolean down)

--- a/client/Android/Studio/freeRDPCore/src/main/java/com/freerdp/freerdpcore/presentation/SessionView.java
+++ b/client/Android/Studio/freeRDPCore/src/main/java/com/freerdp/freerdpcore/presentation/SessionView.java
@@ -295,6 +295,8 @@ public class SessionView extends View
 	// Handle all physical mouse buttons here; finger taps come via onSingleTapUp.
 	@Override public boolean onGenericMotionEvent(MotionEvent event)
 	{
+		if (!event.isFromSource(InputDevice.SOURCE_MOUSE))
+			return false;
 
 		int action = event.getActionMasked();
 
@@ -321,9 +323,23 @@ public class SessionView extends View
 						sessionViewListener.onSessionViewBeginTouch();
 					sessionViewListener.onSessionViewRightTouch(x, y, down);
 					return true;
+				case MotionEvent.BUTTON_TERTIARY:
+					sessionViewListener.onSessionViewMiddleTouch(x, y, down);
+					return true;
 				default:
-					break;
+					return true; // consume unknown buttons silently
 			}
+		}
+
+		if (action == MotionEvent.ACTION_SCROLL)
+		{
+			float vScroll = event.getAxisValue(MotionEvent.AXIS_VSCROLL);
+			float hScroll = event.getAxisValue(MotionEvent.AXIS_HSCROLL);
+			if (vScroll != 0)
+				sessionViewListener.onSessionViewScroll(vScroll > 0);
+			if (hScroll != 0)
+				sessionViewListener.onSessionViewHScroll(hScroll > 0);
+			return true;
 		}
 
 		return false;
@@ -337,6 +353,8 @@ public class SessionView extends View
 
 		void onSessionViewLeftTouch(int x, int y, boolean down);
 
+		void onSessionViewMiddleTouch(int x, int y, boolean down);
+
 		void onSessionViewRightTouch(int x, int y, boolean down);
 
 		void onSessionViewMove(int x, int y);
@@ -344,6 +362,8 @@ public class SessionView extends View
 		void onSessionViewMouseMove(int x, int y);
 
 		void onSessionViewScroll(boolean down);
+
+		void onSessionViewHScroll(boolean right);
 	}
 
 	private class SessionGestureListener extends GestureDetector.SimpleOnGestureListener

--- a/client/Android/Studio/freeRDPCore/src/main/java/com/freerdp/freerdpcore/presentation/SessionView.java
+++ b/client/Android/Studio/freeRDPCore/src/main/java/com/freerdp/freerdpcore/presentation/SessionView.java
@@ -18,9 +18,11 @@ import android.graphics.Matrix;
 import android.graphics.Rect;
 import android.graphics.RectF;
 import android.graphics.drawable.BitmapDrawable;
+import android.os.Build;
 import android.text.InputType;
 import android.util.AttributeSet;
 import android.util.Log;
+import android.view.InputDevice;
 import android.view.KeyEvent;
 import android.view.MotionEvent;
 import android.view.ScaleGestureDetector;
@@ -269,9 +271,62 @@ public class SessionView extends View
 
 	@Override public boolean onTouchEvent(MotionEvent event)
 	{
+		// Physical mouse events: bypass gesture detector entirely.
+		// Buttons are handled in onGenericMotionEvent; hover moves in onHoverEvent.
+		// Only ACTION_MOVE with a button held (drag) needs handling here.
+		if (event.isFromSource(InputDevice.SOURCE_MOUSE))
+		{
+			int action = event.getActionMasked();
+			if (action == MotionEvent.ACTION_MOVE && event.getButtonState() != 0)
+			{
+				MotionEvent mapped = mapTouchEvent(event);
+				sessionViewListener.onSessionViewMouseMove((int)mapped.getX(), (int)mapped.getY());
+				mapped.recycle();
+				return true;
+			}
+			return true;
+		}
+
 		boolean res = gestureDetector.onTouchEvent(event);
 		res |= doubleGestureDetector.onTouchEvent(event);
 		return res;
+	}
+
+	// Handle all physical mouse buttons here; finger taps come via onSingleTapUp.
+	@Override public boolean onGenericMotionEvent(MotionEvent event)
+	{
+
+		int action = event.getActionMasked();
+
+		if (action == MotionEvent.ACTION_BUTTON_PRESS ||
+		    action == MotionEvent.ACTION_BUTTON_RELEASE)
+		{
+			boolean down = action == MotionEvent.ACTION_BUTTON_PRESS;
+			MotionEvent mapped = mapTouchEvent(event);
+			int x = (int)mapped.getX();
+			int y = (int)mapped.getY();
+			mapped.recycle();
+
+			switch (event.getActionButton())
+			{
+				case MotionEvent.BUTTON_PRIMARY:
+					if (down)
+						sessionViewListener.onSessionViewBeginTouch();
+					sessionViewListener.onSessionViewLeftTouch(x, y, down);
+					if (!down)
+						sessionViewListener.onSessionViewEndTouch();
+					return true;
+				case MotionEvent.BUTTON_SECONDARY:
+					if (down)
+						sessionViewListener.onSessionViewBeginTouch();
+					sessionViewListener.onSessionViewRightTouch(x, y, down);
+					return true;
+				default:
+					break;
+			}
+		}
+
+		return false;
 	}
 
 	public interface SessionViewListener
@@ -285,6 +340,8 @@ public class SessionView extends View
 		void onSessionViewRightTouch(int x, int y, boolean down);
 
 		void onSessionViewMove(int x, int y);
+
+		void onSessionViewMouseMove(int x, int y);
 
 		void onSessionViewScroll(boolean down);
 	}
@@ -348,28 +405,18 @@ public class SessionView extends View
 
 		public boolean onSingleTapUp(MotionEvent e)
 		{
-			// send single click
+			// Physical mouse buttons are handled via ACTION_BUTTON_PRESS in onGenericMotionEvent.
+			// If buttonState is non-zero this event came from a physical mouse button
+			if (e.getButtonState() != 0)
+				return false;
+
+			// Finger touch -> left click
 			MotionEvent mappedEvent = mapTouchEvent(e);
 			sessionViewListener.onSessionViewBeginTouch();
-			switch (e.getButtonState())
-			{
-				case MotionEvent.BUTTON_PRIMARY:
-					sessionViewListener.onSessionViewLeftTouch((int)mappedEvent.getX(),
-					                                           (int)mappedEvent.getY(), true);
-					sessionViewListener.onSessionViewLeftTouch((int)mappedEvent.getX(),
-					                                           (int)mappedEvent.getY(), false);
-					break;
-				case MotionEvent.BUTTON_SECONDARY:
-					sessionViewListener.onSessionViewRightTouch((int)mappedEvent.getX(),
-					                                            (int)mappedEvent.getY(), true);
-					sessionViewListener.onSessionViewRightTouch((int)mappedEvent.getX(),
-					                                            (int)mappedEvent.getY(), false);
-					sessionViewListener.onSessionViewLeftTouch((int)mappedEvent.getX(),
-					                                           (int)mappedEvent.getY(), true);
-					sessionViewListener.onSessionViewLeftTouch((int)mappedEvent.getX(),
-					                                           (int)mappedEvent.getY(), false);
-					break;
-			}
+			sessionViewListener.onSessionViewLeftTouch((int)mappedEvent.getX(),
+			                                           (int)mappedEvent.getY(), true);
+			sessionViewListener.onSessionViewLeftTouch((int)mappedEvent.getX(),
+			                                           (int)mappedEvent.getY(), false);
 			sessionViewListener.onSessionViewEndTouch();
 			return true;
 		}

--- a/client/Android/Studio/freeRDPCore/src/main/java/com/freerdp/freerdpcore/utils/Mouse.java
+++ b/client/Android/Studio/freeRDPCore/src/main/java/com/freerdp/freerdpcore/utils/Mouse.java
@@ -19,12 +19,14 @@ public class Mouse
 
 	private final static int PTRFLAGS_LBUTTON = 0x1000;
 	private final static int PTRFLAGS_RBUTTON = 0x2000;
+	private final static int PTRFLAGS_MBUTTON = 0x4000;
 
 	private final static int PTRFLAGS_DOWN = 0x8000;
 	private final static int PTRFLAGS_MOVE = 0x0800;
 
 	private final static int PTRFLAGS_WHEEL = 0x0200;
 	private final static int PTRFLAGS_WHEEL_NEGATIVE = 0x0100;
+	private final static int PTRFLAGS_HWHEEL = 0x0400;
 
 	public static int getLeftButtonEvent(Context context, boolean down)
 	{
@@ -40,6 +42,11 @@ public class Mouse
 			return (PTRFLAGS_LBUTTON | (down ? PTRFLAGS_DOWN : 0));
 		else
 			return (PTRFLAGS_RBUTTON | (down ? PTRFLAGS_DOWN : 0));
+	}
+
+	public static int getMiddleButtonEvent(boolean down)
+	{
+		return (PTRFLAGS_MBUTTON | (down ? PTRFLAGS_DOWN : 0));
 	}
 
 	public static int getMoveEvent()
@@ -59,6 +66,20 @@ public class Mouse
 			flags |= (PTRFLAGS_WHEEL_NEGATIVE | 0x0088);
 		else
 			flags |= 0x0078;
+		return flags;
+	}
+
+	public static int getHScrollEvent(Context context, boolean right)
+	{
+		int flags = PTRFLAGS_HWHEEL;
+
+		if (ApplicationSettingsActivity.getInvertScrolling(context))
+			right = !right;
+
+		if (right)
+			flags |= 0x0078;
+		else
+			flags |= (PTRFLAGS_WHEEL_NEGATIVE | 0x0088);
 		return flags;
 	}
 }

